kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,672 @@
+-- vim: ts=4 sts=4 sw=4 et:
+
+local utils = require("kong.resty.dns_client.utils")
+local mlcache = require("kong.resty.mlcache")
+local resolver = require("resty.dns.resolver")
+
+local now = ngx.now
+local log = ngx.log
+local ERR = ngx.ERR
+local WARN = ngx.WARN
+local ALERT = ngx.ALERT
+local cjson_encode = require("cjson.safe").encode
+local math_min = math.min
+local timer_at = ngx.timer.at
+local table_insert = table.insert
+local ipv6_bracket = utils.ipv6_bracket
+
+local req_dyn_hook_run_hooks = require("kong.dynamic_hook").run_hooks
+
+-- Constants and default values
+local DEFAULT_ERROR_TTL = 1     -- unit: second
+local DEFAULT_STALE_TTL = 4
+local DEFAULT_EMPTY_TTL = 30
+
+local DEFAULT_ORDER = { "LAST", "SRV", "A", "AAAA", "CNAME" }
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local valid_type_names = {
+    SRV     = TYPE_SRV,
+    A       = TYPE_A,
+    AAAA    = TYPE_AAAA,
+    CNAME   = TYPE_CNAME,
+    LAST    = TYPE_LAST,
+}
+
+local typstrs = {
+    [TYPE_SRV]      = "SRV",
+    [TYPE_A]        = "A",
+    [TYPE_AAAA]     = "AAAA",
+    [TYPE_CNAME]    = "CNAME",
+}
+
+local hitstrs = {
+    [1] = "hit_lru",
+    [2] = "hit_shm",
+}
+
+local errstrs = {     -- client specific errors
+    [100] = "cache only lookup failed",
+    [101] = "no available records",
+}
+
+local EMPTY_ANSWERS = { errcode = 3, errstr = "name error" }
+
+
+-- APIs
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+    if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+        _M[k] = v
+    end
+end
+_M.TYPE_LAST = -1
+
+
+local tries_mt = { __tostring = cjson_encode }
```

Comment:
For reviewers, this __tostring method works with the following @tries table (also in this file)

```
-- insert the error message for this failed attempt.
   table_insert(tries, { name .. ":" .. typstrs[qtype], err })
```

Its example error log of Tries like:

```
2024/02/28 16:35:23 [error] 77335#0: *1336 [kong] init.lua:405 [aws-lambda] Lambda:invoke() failed to connect to 'https://lambda.ab-cdef-1.amazonaws.com:10001': 
[cosocket] DNS resolution failed: dns server error: 3 name error. 
Tried: [["lambda.ab-cdef-1.amazonaws.com:SRV","dns server error: 3 name error"],
["lambda.ab-cdef-1.amazonaws.com:A","dns server error: 3 name error"],
["lambda.ab-cdef-1.amazonaws.com:CNAME","dns server error: 3 name error"]], 
client: 127.0.0.1, server: kong, request: "GET /get?key1=some_value1 HTTP/1.1", host: "lambda15.test", request_id: "3333b77722d0bc23a44fad31953ccf0b"
```


---

### another method 

And I also have another implementation with more source lines but it has a bit performance improve during running time

```
local tries_mt = {
    __tostring = function(self)
        local t, i = {}, 0
        for _, e in ipairs(self) do
            local name, typ, err = e[1], typstrs[e[2]], e[3]
            t[i] = '", "'
            t[i+1] = name
            t[i+2] = ":"
            t[i+3] = typ
            t[i+4] = " - "
            t[i+5] = err
            i = i + 6
        end
        return '["' .. table.concat(t) .. '"]'
    end,
}

-- insert the error message for this failed attempt.
--  no string concatenation here like: { "<name>:<type_str>", err }
   table_insert(tries, { name, qtype, err })
```